src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/LspProjectEngineFactoryProvider.cs

```diff
@@ -0,0 +1,70 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System;
+using System.Diagnostics.CodeAnalysis;
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.AspNetCore.Razor.ProjectEngineHost;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer;
+
+/// <summary>
+/// In the language server, we provide an <see cref="IProjectEngineFactoryProvider"/> that wraps
+/// <see cref="ProjectEngineFactories.DefaultProvider"/> and configure every <see cref="RazorProjectEngine"/>
+/// with the current code-gen options.
+/// </summary>
+internal sealed class LspProjectEngineFactoryProvider(IOptionsMonitor<RazorLSPOptions> optionsMonitor) : IProjectEngineFactoryProvider
+{
+    [return: NotNullIfNotNull(nameof(fallbackFactory))]
```

Comment:
It's part of the interface contract. IMO, it can be a bit confusing to change nullability attributes on an interface implementation or override. That's not a hard and fast rule, but I always feel a little oogy if there are different nullability semantics when calling through a contract vs. calling through the same method on the implementation.

Reply:
That said, I think the contract is weird. I'm going to take a deeper look and see if "fallback factory" is something that's really even needed. I can imagine that we just always say that we return an empty project engine factory or even throw.

Reply:
I think I ended up seeing this on the interface, and should have deleted the comment here. Fine with this as is, any further clean up can always be a follow up PR. Don't let me randomised you too much :)

Reply:
OK, I revisited the `IProjectEngineFactory` and `IProjectEngineFactoryProvider` contracts and they're much simpler now.